teps/0076-array-result-types.md

```diff
@@ -0,0 +1,620 @@
+   could index into the result. (Note that the existing gcloud task provides no results at all.)
+
+   Example gcloud output:
+
+   ```
+   # gcloud container clusters list
+   NAME        LOCATION     MASTER_VERSION   MASTER_IP     MACHINE_TYPE   NODE_VERSION         NUM_NODES  STATUS
+   eucalyptus  us-central1  1.18.20-gke.501  1.1.1.1  n1-standard-4  1.16.15-gke.6000 **  3          RUNNING
+   ```
+
+   Using this as a result would require supporting nested arrays, which this TEP is not proposing but would be a logical
+   next step (e.g. `$tasks.gcloud.results.[1][1]` if you wanted the location of the first cluster OR even to get the
+   entire line and provide it as an array param with `$tasks.gcloud.results[1]`).
+
+## Requirements
+
+* Must be possible for a result array to be empty
+* Must be possible to use the array results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has an array param
+* Must be possible to use one specific value in an array result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value at the specified index, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+* This must be support for all uses of results, including
+  [pipeline results](https://github.com/tektoncd/pipeline/blob/main/docs/pipelines.md#emitting-results-from-a-pipeline)
+
+## Proposal
+
+Currently we treat the contents of
+[a Task's result file(s)](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results)
+as a string. The file is located at `$(results.resultName.path)` which expands to `/tekton/results/resultName`.
+In this proposal, the results section of the Task will inform the controller about whether it should interpret the
+content of the result as [json](#why-json) or as a string:
+  * If the type indicates that the result is an array, the controller will interpret the result as json and expect an
+    array. If the content is not json or not an array, the TaskRun will fail
+  * If the type indicates that the result is a string, the controller will simply use the result as is (i.e. the
+    same logic as today)
+
+Pros:
+  * Completely backwards compatible
+
+Cons:
+  * Additional complexity in that the results file could contain two different kinds of content
+
+This feature would be considered alpha and would be
+[gated by the alpha flag](https://github.com/tektoncd/pipeline/blob/main/docs/install.md#alpha-features).
+
+### Why json?
+
+This proposal suggests using `json` as the format for more complex types like arrays, and in
+[TEP-0075](https://github.com/tektoncd/community/pull/479), dictionaries (technically "objects"). Possible other formats
+we could use are:
+
+* `yaml` (i dont think there is a syntax like jsonpath available for yaml? i.e. it'd be harder to expose via variable
```

Comment:
yaml ends up being in-memory, a memory structure that's very similar to json, so you should be able to use jsonpath / jsonpointer etc on a yaml document.